lib/core/events/emitter.js

```diff
@@ -0,0 +1,254 @@
+    super();
+    this.superEmit = super.emit;
+    this.pipeRunner = new PipeRunner(maxConcurrentPipes, pipesBufferSize);
+
+    // register caches
+    this.pluginPipes = new Map();
+    this.corePipes = new Map();
+    this.coreAnswerers = new Map();
+  }
+
+  /**
+   * Registers a core method on a pipe
+   * Note: core methods cannot listen to wildcarded events, only exact matching
+   * works here.
+   *
+   * @param  {String}   event
+   * @param  {Function} fn
+   */
+  onPipe (event, fn) {
+    assert(typeof fn === 'function', `Cannot listen to pipe event ${event}: "${fn}" is not a function`);
+
+    if (!this.corePipes.has(event)) {
+      this.corePipes.set(event, []);
+    }
+
+    this.corePipes.get(event).push(fn);
+  }
+
+  /**
+   * Registers a core 'ask' event answerer
+   * There can only be 0 or 1 answerer per ask event.
+   *
+   * @param  {String}   event
+   * @param  {Function} fn
+   */
+  onAsk (event, fn) {
+    assert(typeof fn === 'function', `Cannot listen to ask event "${event}": "${fn}" is not a function`);
+    assert(!this.coreAnswerers.has(event), `Cannot add a listener to the ask event "${event}": event has already an answerer`);
+
+    this.coreAnswerers.set(event, fn);
+  }
+
+  /**
+   * Emits an event and all its wildcarded versions
+   *
+   * @warning Critical section of code
+   *
+   * @param  {string} event
+   * @param  {*} data
+   */
+  emit (event, data) {
+    const events = getWildcardEvents(event);
+    debug('Triggering event "%s" with data: %o', event, data);
+
+    for (let i = 0; i < events.length; i++) {
+      super.emit(events[i], data);
+    }
+  }
+
+  /**
+   * Emits a pipe event, which triggers the following, in that order:
+   * 1. Plugin pipes are invoked one after another (waterfall). Each plugin must
+   *    resolve the pipe (with a callback or a promise) with a similar payload
+   *    than the one received
+   * 2. Core pipes are invoked in parallel. They are awaited for (promises-only)
+   *    but their responses are neither evaluated nor used
+   * 3. Hooks are invoked in parallel. They are not awaited.
+   *
+   * Accepts a callback argument (to be used by pipes invoked before the funnel
+   * overload-protection mechanism). If a callback is provided, this method
+   * doesn't return a promise.
+   *
+   * @warning Critical section of code
+   *
+   * @param  {string} event
+   * @param  {*} payload
+   * @return {Promise.<*>|null}
+   */
+  pipe (event, ...payload) {
+    debug('Triggering pipe "%s" with payload: %o', event, payload);
+
+    let callback = null;
+
+    // safe: a pipe's payload can never contain functions
+    if (payload.length > 0 && typeof payload[payload.length-1] === 'function') {
+      callback = payload.pop();
+    }
+
+    const events = getWildcardEvents(event);
+    const funcs = [];
+
+    for (let i = 0; i < events.length; i++) {
+      const targets = this.pluginPipes.get(events[i]);
+
+      if (targets) {
+        funcs.push(...targets);
```

Comment:
Since statiscally, event arrays should be small, I think you're right and I'll change this.

But beware, as the fastest method depends on the expected size of the array to inject.

Tiny array:

```
$ NODE_ENV=production node bench2 1
for..of loop x 16,251,316 ops/sec ±0.69% (93 runs sampled)
spread x 14,688,402 ops/sec ±0.38% (97 runs sampled)
array concat x 5,311,926 ops/sec ±0.40% (92 runs sampled)
for(;;) loop x 18,455,715 ops/sec ±0.86% (95 runs sampled)
Fastest is for(;;) loop
```

Small array:
```
$ NODE_ENV=production node bench2 '12345'
for..of loop x 14,030,483 ops/sec ±1.33% (94 runs sampled)
spread x 12,308,696 ops/sec ±1.82% (96 runs sampled)
array concat x 5,245,292 ops/sec ±0.97% (95 runs sampled)
for(;;) loop x 16,371,162 ops/sec ±0.43% (95 runs sampled)
Fastest is for(;;) loop
```
Large array:
```
$ NODE_ENV=production node bench2 'this is a very, very, very long string, because size does matter'
for..of loop x 2,768,369 ops/sec ±0.66% (91 runs sampled)
spread x 3,578,644 ops/sec ±0.37% (95 runs sampled)
array concat x 4,261,540 ops/sec ±0.95% (97 runs sampled)
for(;;) loop x 2,950,494 ops/sec ±0.25% (94 runs sampled)
Fastest is array concat
```


<details><summary>Benchmark code</summary>
const { Suite } = require('benchmark');
const suite = new Suite();

const str = process.argv.length > 1
  ? process.argv[2]
  : 'can i haz cheezburgers? (old meme)';

const buf = str.split('');
const arr = 'foobar'.split('');

suite
  .add('for..of loop', function() {
    const _arr = Array.from(arr);

    for (const f of buf) {
      _arr.push(f);
    }
  })
  .add('spread', async function() {
    const _arr = Array.from(arr);

    _arr.push(...buf);
  })
  .add('array concat', async function () {
    let _arr = Array.from(arr);

    _arr = _arr.concat(buf);
  })
  .add('for(;;) loop', function () {
    const _arr = Array.from(arr);

    for (let i = 0; i < buf.length; i++) {
      _arr.push(buf[i]);
    }
  })
  .on('cycle', function(event) {
    console.log(String(event.target));
  })
  .on('complete', function() {
    console.log('Fastest is ' + this.filter('fastest').map('name'));
  })
  .run();
</details>